Upgraded enemies outcome

IndoorKat opened the pull request:
Hi everyone, I've been thinking about this outcome for a while so decided to try making it.
Right now it's probably not balanced, it turns:
Ukko -> Super ukko
Firebug -> Big firebug
Stendari -> better stendari (blue flame one)
Steve -> big steve
Small worm -> meat realm worm
Medium worm -> skull worm
Big worm -> hell worm
Hound -> big hound
Acid shooter -> strong acid shooter
Slime shooter -> laser shooter
Shotgunner -> hell shotgunner
Tnt thrower -> hell tnt thrower
Sniper -> hell sniper
Blue fungus dude -> bigger pink fungus dude
Tank -> laser tank
Necrobot -> super necrobot
Robocop -> better robocop

From a balance point of view I see a couple ways forward:
1. Make the effect not time limited and remove the deadlier ones or randomly select some of them for a worse NG+ enemies feel
2. Make the effect time limited and keep deadly upgrades for a steve gang feel

Another thing to note is that biomes vary quite a bit in how many of these 'upgradable' enemies exist e.g. the mines are filled with them while the coal pits only have a couple.

An issue:
The enemies which are spawned per biome (i.e. the vault ukko) have their name as an empty string, so they don't get affected, I'd appreciate some advice on how to approach fixing this.

Also if anyone has any other enemies that they'd view as 'upgrades' to another lmk

IndoorKat committed:
Outcome: upgraded enemies

IndoorKat committed:
remove vote history from my version

IndoorKat committed:
Remove debugging stuff

Miczu commented:
So first could you clean up obs.txt and remove changes from "package-lock.json"? Those are not necessary for your PR.

For me to test your changes (and Chase to download the update), you will need to create new entries to manifest.json.
That file stores path to the files and sha256 hash of the file. Simple to generate the hash is just to drag the file to site like https://www.strerr.com/en/sha256_file.html
It is common for github to add/remove/change some small things in a file, so downloading file from github and using that file for sha256 is most reliable.

Without testing... I wonder if you can make the effect not instant, but like delayed effect that upgrades one enemy at a time. So player cans see "F, sniper is upgrading, RUN".

Maybe gamify it with something like damaging transforming enemy interrupts him or slows down the transformation.

Maybe enemy needs to have line of sight with player?

IndoorKat committed:
Remove obs

IndoorKat committed:
Upgrade delay, upgrade icon, only begin when in player line of sight

IndoorKat committed:
remove obs again

IndoorKat commented:
Alright, added a line of sight check, a delay, and an icon, I'd still like to fix enemies from specific biomes not transforming.

Miczu commented:
Branch works after hammering out sha256 issues
<img width="698" height="388" alt="image" src="https://github.com/user-attachments/assets/839b08ae-1c61-442d-a202-266198d079fc" />

Miczu commented:
@IndoorKat the effect doesn't stack correctly. You should find existing effect and add duration. Feel free to reference other effects with duration
<img width="149" height="213" alt="image" src="https://github.com/user-attachments/assets/bf84a3a0-d00f-4f0d-bd9b-12194d199da5" />

Miczu commented:
Hissi sniper and hissi shotgun are bit strong and player has little time to react and enemies transform out of view.
- Reduce the effect radius
- Add maybe 2 seconds to the transformation time
- would be nice to manipulate the some enemies hp to not be full meat biome mini bosses (not sure what would be good value, maybe start with 1/2 ?)

Teleport mages are used for speedruns... maybe you have fun mage upgrade in mind? 😸 
Jungle biome is upgrade poor - maybe some spiders could use an upgrade?

Vault variant of ukko, tank gattling, sniper and robocop doesn't transformt. I suspect it is called differently in game
<img width="273" height="305" alt="image" src="https://github.com/user-attachments/assets/10cd5ab9-67f9-411f-814b-777e2e3db11d" />

Temple ukko doesn't transform.
Temple critically lacks any upgrades - dead vote. Maybe you can transform big maggots and harmless scorpions?

IndoorKat commented:
Yeah the later enemy variants (i.e. vault ukko, vault robocop, etc) have no entity name which is how I’m checking what to upgrade, I think I can switch over to getting the filepath to fix this. I’ll make those balance changes and make later more deadly too.

IndoorKat committed:
Changes

IndoorKat committed:
changes

IndoorKat commented:
Fixed stacking, fixed enemy name issue (using filepaths now), capped max HP for enemies at 50 for the first zone (player should have decent damage by coal pits), decreased detection radius, increased transformation time, added the following new transformations:
spirits -> diminution spirit
mages -> poly mage
green spiders in jungle -> luuki lair spiders
scorpion -> mecha kolmi (spice up the temple a bit, they are fairly rare and can be easily one shot so I think it's balanced and funny)
small tentacler -> big tentacler
hiisii healer -> hiisii shielder

IndoorKat commented:
currently a bug with multiple enemies being spawned per enemy sometimes, looking into it

IndoorKat committed:
Fix multiple enemies spawning

IndoorKat commented:
fixed

Miczu commented:
Sounds like spicy updates, I'll try test after work, but I may be distracted by silksong 🕷️

Miczu commented:
@IndoorKat your sha256 hashes are consistantly incorrect 
<img width="990" height="231" alt="image" src="https://github.com/user-attachments/assets/6078aba6-7555-49fd-9e38-9002bcb55454" />
Common issue is that github normalizes file format, so your local files could be not good for generating hash.

IndoorKat committed:
Hopefully fix manifest.json

IndoorKat commented:
Might be fixed now, I downloaded the files off github, chucked them in the sha256 website and then updated manifest

IndoorKat commented:
Looks like the new one matches for outcomes/upgraded_enemies.lua from your screenshot at least.

Miczu commented:
Should be good for my testing later after work 👍 
<img width="653" height="386" alt="image" src="https://github.com/user-attachments/assets/80263d52-3d56-47e8-8dea-6d159e1cafdb" />

Miczu commented:
Looks good, should be add some spicy moments. Hope he builds dps wand or is lucky with his path going down.

Miczu reviewed (approved):
Works on my machine
